File: core/network/src/main/java/app/muplay/network/SubsonicAuth.java
package app.muplay.network;

import app.muplay.model.SubsonicCredentials;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Supplier;
import javax.annotation.Nonnull;

/**
 * Builds Subsonic authentication query parameters.
 *
 * <p>Navidrome 0.63.2 supports only {@code p=} and {@code t}+{@code s}; the
 * OpenSubsonic {@code apiKeyAuthentication} extension is not implemented. We use
 * token auth and never put the password on the wire.
 */
public final class SubsonicAuth {

  /** Identifies this client. Must not be "DSub" or "SubMusic" — Navidrome
   * strips the entire OpenSubsonic field block for those. */
  public static final String CLIENT_NAME = "MuPlay";

  public static final String PROTOCOL_VERSION = "1.16.1";

  private static final String SALT_ALPHABET =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
  private static final int SALT_LENGTH = 12;

  private static final SecureRandom RANDOM = new SecureRandom();

  @Nonnull
  public static Supplier<String> randomSaltSupplier() {
    return () -> {
      StringBuilder sb = new StringBuilder(SALT_LENGTH);
      for (int i = 0; i < SALT_LENGTH; i++) {
        sb.append(SALT_ALPHABET.charAt(RANDOM.nextInt(SALT_ALPHABET.length())));
      }
      return sb.toString();
    };
  }

  /** {@code t = md5(password + salt)}, lowercase hex. */
  @Nonnull
  public static String token(@Nonnull String password, @Nonnull String salt) {
    try {
      MessageDigest md5 = MessageDigest.getInstance("MD5");
      byte[] digest = md5.digest((password + salt).getBytes(StandardCharsets.UTF_8));
      StringBuilder sb = new StringBuilder(32);
      for (byte b : digest) {
        sb.append(Character.forDigit((b >> 4) & 0xF, 16));
        sb.append(Character.forDigit(b & 0xF, 16));
      }
      return sb.toString();
    } catch (NoSuchAlgorithmException e) {
      throw new IllegalStateException("MD5 is required by the Java platform", e);
    }
  }

  @Nonnull
  public Map<String, String> authParams(
      @Nonnull SubsonicCredentials credentials, @Nonnull Supplier<String> saltSupplier) {
    String salt = saltSupplier.get();
    Map<String, String> params = new LinkedHashMap<>();
    params.put("u", credentials.username());
    params.put("t", token(credentials.password(), salt));
    params.put("s", salt);
    params.put("v", PROTOCOL_VERSION);
    params.put("c", CLIENT_NAME);
    params.put("f", "json");
    return Map.copyOf(params);
  }
}

File: core/network/src/main/kotlin/app/muplay/network/SubsonicAuth.kt
package app.muplay.network

import app.muplay.model.SubsonicCredentials
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

/**
 * Subsonic API token authentication (https://www.subsonic.org/pages/api.jsp#authentication).
 *
 * The salt is a caller-supplied parameter rather than something generated internally, which keeps
 * [token] and [authParams] pure and deterministic for tests. Production wiring must generate a
 * fresh salt per request with `SecureRandom` — never cache or reuse one.
 */
object SubsonicAuth {

  /**
   * The client identifier sent as the `c` parameter. Not cosmetic: Navidrome's
   * `Subsonic.LegacyClients` defaults to `DSub` and `MinimalClients` to `SubMusic`, and a client
   * whose `c` value matches either of those has the entire OpenSubsonic field block stripped from
   * every response.
   */
  const val CLIENT_NAME: String = "MuPlay"

  /** The Subsonic API protocol version this client targets, sent as the `v` parameter. */
  const val PROTOCOL_VERSION: String = "1.16.1"

  /**
   * Computes the Subsonic token: `hex(md5(utf8Bytes(password) + utf8Bytes(salt)))`, lower-cased.
   *
   * Hex-encodes byte by byte rather than via `BigInteger`/`Integer.toHexString` over the whole
   * digest, both of which silently drop a leading zero byte — a bug that only shows up for
   * roughly one salt in sixteen.
   */
  fun token(password: String, salt: String): String {
    val digest = MessageDigest.getInstance("MD5")
      .digest((password + salt).toByteArray(StandardCharsets.UTF_8))
    return digest.joinToString(separator = "") { byte -> "%02x".format(byte) }
  }

  /**
   * Builds the Subsonic authentication query parameters for [credentials] using [salt]. Never
   * includes the plaintext password: only the salted [token] (`t`) and the [salt] itself (`s`)
   * are emitted, per the Subsonic token-authentication scheme.
   */
  fun authParams(credentials: SubsonicCredentials, salt: String): Map<String, String> =
    mapOf(
      "u" to credentials.username,
      "t" to token(credentials.password, salt),
      "s" to salt,
      "v" to PROTOCOL_VERSION,
      "c" to CLIENT_NAME,
      "f" to "json",
    )
}
